### ISSUE TITLE
[Feature Request] Online (web) & PDF invoice template editor

### ISSUE THREAD

**Issue by ririko5834** [InvoicePlane/InvoicePlane-v2#401]
It would be nice to add a template editor for online (web public) invoice templates. So you can build your own invoice template (using some HTML editor) with template tags (placeholders) without the need of coding PHP skills.

**Comment by clockwiseq** [comment]
This is an enhancement that I recommended a couple of years ago.  Not sure what direction we go in with the v1 vs v2 _(branch the code and start over)_ or if we continue on with the current code base and enhance it.  Either way, this is one of my favorite requests as I have always felt that it would increase the user-base because business owners wouldn't have to depend on a PHP developer to create their own invoice templates.

**Comment by ririko5834** [comment]
The same thing should exist for PDF invoice templates.

With this, you could change the appearance of the templates, for example, make that custom client fields will be displayed in the templates

**Comment by clockwiseq** [comment]
Yeah, that was what I meant above.  The template editor would allow you to modify and store the templates in the database and have a visual editor for ease-of-use.  Those templates would be for the three (3) areas an invoice was displayed _(web, email, PDF)_

**Comment by OPGRBILLINGDEV** [comment]
Is underlyning technology selected for template editor ? What about pure CSS ?

Certain open source solutions exist in Git, such as rukavina's>
https://github.com/topics/css-editor

...and could be merged/tweeked/forked into/with 1.6.x or v2.

Has also features like - background image option for web nd PDF, which we do not need to extra-code into IP, just add editor which has such features.

Paralell CSS editor with wysiwig window, exist also in open-source CSS editor from sauravhathi>
https://github.com/sauravhathi/live-html-editor

**Comment by nielsdrost7** [comment]
@coderabbitai attach to issues regarding the report builder

**Comment by nielsdrost7** [comment]
## Planning Note — v2 Implementation

**Story:** As a company admin with no PHP knowledge, I want a visual drag-and-drop editor for invoice and quote PDF templates, so that I can customise the layout (logo position, colour scheme, field order) without editing Blade files.

**v2 context:** v2 generates PDFs from Blade templates using `barryvdh/laravel-dompdf`. A full drag-and-drop template editor is a substantial feature. The pragmatic v2 approach is a **CSS/variable override system** plus a **template selector** — giving 80% of the value at 20% of the complexity.

**Acceptance Criteria:**
- Company Settings → Invoice Templates has a list of available PDF templates (at least: `default`, `classic`, `modern`).
- Each template has configurable variables: primary colour, logo position (left/right), show/hide fields (e.g. hide IBAN).
- The selected template and its variables are stored per company.
- A preview of the rendered PDF is shown (using a sample invoice) when the admin changes template settings.
- The existing Blade templates are kept for advanced users who want to customise them directly.

**Implementation:**
- `Company` model gains `pdf_template` (string, default `default`) and `pdf_template_options` (JSON).
- Template Blade files: `Modules/Core/resources/Templates/Views/templates/invoices/{template}.blade.php`.
- Settings page: Filament `Select` for template choice + dynamic `ColorPicker` / `Toggle` fields driven by the selected template's config.
- PDF generation: `InvoiceService::generatePdf($invoice)` reads `$invoice->company->pdf_template` and passes `pdf_template_options` as Blade data.
- Preview endpoint: `GET /{tenant}/invoice-template-preview` — renders a mock invoice with current settings.

**Post-v2 scope:** A full drag-and-drop block editor (like Unlayer/GrapeJS) embedded in the settings page is a significant standalone feature that deserves its own tracking in the report builder epic (#130).

**Test Scenarios (PHPUnit):**
```php
// PDF uses company's selected template
/* Arrange */ $this->company->update(['pdf_template' => 'modern']);
              $invoice = Invoice::factory()->for($this->company)->create();
/* Act */     $html = app(InvoiceService::class)->renderPdfHtml($invoice);
/* Assert */  $this->assertStringContainsString('modern-template', $html);
```
```php
// Template options (colour) are passed to blade
/* Arrange */ $this->company->update(['pdf_template_options' => ['primary_color' => '#FF5733']]);
              $invoice = Invoice::factory()->for($this->company)->create();
/* Act */     $html = app(InvoiceService::class)->renderPdfHtml($invoice);
/* Assert */  $this->assertStringContainsString('#FF5733', $html);
```

**Comment by nielsdrost7** [comment]
Update 2026-07-04 — this ask is being delivered through the **Report Builder** (#130, epic #506) rather than a raw source-code editor. Mapping of what was requested here to what ships:

| Ask in this issue | How it's delivered |
|---|---|
| In-browser editing of invoice PDF templates | Visual builder: five bands (header / group header / details / group footer / footer) composed of drag-and-drop **Bricks** (Company, Client, Items, Totals, …) — no code editing needed |
| Sidebar listing available placeholders | Each brick's config panel exposes its fields as toggles/options (e.g. show VAT id, show phone) — the placeholder list becomes structured configuration |
| Preview with sample data | The builder's **Preview** action renders the template with dummy data |
| Save persists the template | Templates are saved as data files under `storage/app/report_templates/` (we deliver system templates, users clone and customize) |

Why not a CodeMirror/Monaco Blade editor as originally described: letting users edit and persist Blade/PHP source from the browser is a template-injection → remote-code-execution risk (any compromised admin account could run arbitrary code on the server). The brick approach keeps all executable code shipped with the app and reviewed; user input is configuration data only, validated against each brick's schema.

The five sub-tasks that were parented here (#519, #521, #523, #525, #528) have been re-scoped to the builder architecture and reparented to #130. Keeping this issue open until the builder ships an equivalent editing experience; roadmap and acceptance criteria live in #130.

**Comment by nielsdrost7** [comment]
## SMART Plan — Verified

The plan in this issue is **accurate and actionable**. Code review confirms:

- No PDF template editor exists in `Modules/Core/Filament/Admin/Pages/` or anywhere in the codebase.
- Related issues #96 (inline PDF preview on invoice edit page) and #190 (PDF renderer stability) are dependencies — they are not yet implemented but the issue notes they should be stable first.
- The app uses WKHTMLTOPDF for PDF generation (storage/DejaVuSans.ufm.json etc. present), so the renderer is available.
- The plan to use a Livewire component wrapping CodeMirror or Monaco is sound given the tech stack.

**Scope clarification needed**: 
- Should the template be file-based (persisted to `resources/views/invoices/pdf.blade.php`) or database-backed (`templates` table)?
- Should the initial implementation support Invoice templates only, or all document types (quotes, etc.) from day one?

This issue depends on #96 and #190 being stable first. Recommend confirming template storage strategy before implementation.

**Comment by maxcomperatore** [comment]
if you're thinking template-first architecture for the editor, worth looking at how [rubrol](https://github.com/maxcomperatore/rubrol) approaches it - templates are version-controlled files, data comes in via JSON, PDF out. no browser in the loop. the separation of template from renderer makes the editor surface way cleaner to build
